packages/plugin-ext/src/main/browser/plugin-contribution-handler.ts

```diff
@@ -187,6 +187,7 @@ export class PluginContributionHandler {
         if (contributions.taskDefinitions) {
             contributions.taskDefinitions.forEach(def => this.taskDefinitionRegistry.register(def));
         }
+        this.taskDefinitionRegistry.setReady();
```

Comment:
I don't think that it is a good design to allow a client extension to decide when the task extension is ready.

Could you think about another solution maybe with events, i.e. when task definitions are registered, `onDidRegister` event is fired and clients who are interested in new definitions get updated?

Reply:
task-loading requires all task definitions to be registered to function correctly. If task definitions are only **partially** in place, task-loading would still break.

are you describing a solution, where the loaded tasks are updated in memory each and every time the task definition registry receives new data?

Reply:
> are you describing a solution, where the loaded tasks are updated in memory each and every time the task definition registry receives new data?

something like that

The point is that plugin system should not be responsible to say when all tasks are loaded since it does not know. There can be other clients, like Theia extensions which not loaded tasks yet. Also the plugin system can uninstall and install new plugins at any time. How `ready` is going to help with it?

There should be: **synchronous** api to access tasks loaded on the current tick and an event indicating that more tasks are loaded (maybe with an increment) that clients can consider new tasks or reconcile all of them.

Alternatively, you can add something like `TaskContribution` point which is called from the task extension and awaits before initial tasks are loaded, then the plugin system should await initial deployment and only then resolve it. Se such apis like `CommandContribution`, `MenuContribution` and so on for the inspiration.

Reply:
> There should be: synchronous api to access tasks loaded on the current tick and an event indicating that more tasks are loaded (maybe with an increment) that clients can consider new tasks or reconcile all of them.

@akosyakov thank you for the feedback
what i am dealing with in this pull request is the `TaskDefinition`, not tasks.
maybe i can work on the tasks api you described in a separate pull request in near future.

one thing i don't quite understand is the word **synchronous**
some tasks are loaded from files, how are we able to make "accessing tasks" synchronous?

if we looks at the code in vs code, both "get one task" and "get all tasks" are async:

https://github.com/microsoft/vscode/blob/f06011ac164ae4dc8e753a3fe7f9549844d15e35/src/vs/workbench/contrib/tasks/browser/abstractTaskService.ts#L473

https://github.com/microsoft/vscode/blob/f06011ac164ae4dc8e753a3fe7f9549844d15e35/src/vs/workbench/contrib/tasks/browser/abstractTaskService.ts#L501

could you please clarify ? Thanks !

Reply:
just want to clarify that, this pull request is related to `TaskDefinition`, not `Task`. I could implement the Task api as per your suggestion in a separate PR in near future. 

In the latest version, I added an event emitter to the `TaskDefinitionRegistry`, so that all its clients are notified when definitions are added or removed, and thus can adjust how tasks are classified in the memory.